packages/next/src/build/webpack/plugins/slow-module-detection-plugin.ts

```diff
@@ -0,0 +1,248 @@
+
+        this.moduleParents.set(child, parent)
+
+        let parentChildren = this.moduleChildren.get(parent)
+        if (!parentChildren) {
+          parentChildren = new Map()
+          this.moduleChildren.set(parent, parentChildren)
+        }
+        parentChildren.set(getModuleIdentifier(child), child)
+      }
+    }
+
+    this.isFinalized = true
+  }
+
+  generateReport(compilation: Compilation) {
+    if (!this.isFinalized) {
+      this.prepareReport(compilation)
+    }
+
+    // Find root modules (those with no parents)
+    const rootModules = [...this.modules.values()].filter(
+      (node) => !this.moduleParents.has(node)
+    )
+
+    const formatModuleNode = (node: Module, depth: number): string => {
+      const moduleName = getModuleDisplayName(node) || ''
+
+      if (!moduleName) {
+        return formatChildModules(node, depth)
+      }
+
+      const buildInfo = getModuleBuildInfo(node)
+      const buildTimeMs = buildInfo.slowModuleDetectionTiming
+
+      const duration =
+        buildTimeMs && buildTimeMs > 0
+          ? yellow(` (${Math.ceil(buildTimeMs)}ms)`)
+          : ''
+
+      const indentation = ' ' + TreeSymbols.VERTICAL_LINE.repeat(depth)
+      const moduleText = blue(truncatePath(moduleName))
+
+      return (
+        indentation +
+        TreeSymbols.BRANCH +
+        moduleText +
+        duration +
+        '\n' +
+        formatChildModules(node, depth + 1)
+      )
+    }
+
+    const formatChildModules = (node: Module, depth: number): string => {
+      const children = this.moduleChildren.get(node)
+      if (!children) return ''
+
+      return [...children]
+        .map(([_, child]) => formatModuleNode(child, depth))
+        .join('')
+    }
+
+    const report = rootModules.map((root) => formatModuleNode(root, 0)).join('')
+
+    if (report) {
+      console.log(
+        green(
+          `🐌 Detected slow modules while compiling ${this.options.compilerType}:`
+        ) +
+          '\n' +
+          report
+      )
+    }
+  }
+}
+
+export default class SlowModuleDetectionPlugin {
+  constructor(private options: ModuleBuildTimeAnalyzerOptions) {}
+
+  apply = (compiler: Compiler) => {
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      const analyzer = new ModuleBuildTimeAnalyzer(this.options)
+      const moduleBuildStartTimes = new WeakMap<Module, number>()
+
+      compilation.hooks.buildModule.tap(PLUGIN_NAME, (module) => {
+        moduleBuildStartTimes.set(module, performance.now())
+      })
+
+      compilation.hooks.succeedModule.tap(PLUGIN_NAME, (module) => {
+        const startTime = moduleBuildStartTimes.get(module)
+        if (!startTime) {
+          throw new Error(
+            `Invariant (SlowModuleDetectionPlugin): Unable to find the start time for a module build. This is a Next.js internal bug.`
+          )
+        }
+        analyzer.recordModuleBuildTime(module, performance.now() - startTime)
```

Comment:
It will give you some incorrect reports due to the single threaded nature of node.js. if one module compilation is blocking the main thread, all concurrent modules will be reported as slow.

It might also report modules as slower than they are because we are running it concurrently.

Technically you would need to set concurrency to 1 for a good measurement